core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -7355,4 +7465,11 @@ private enum Clause {
     ORDER,
     CURSOR
   }
+
+  /** Which clause is being expanded. */
```

Comment:
GroupByScope calls into the validator like so:

```
    SqlNode expanded =
        validator.extendedExpand(expr, this, select,
            SqlValidatorImpl.Clause.GROUP_BY);
```

The enum has to be public. 

Also we only expand GROUP BY , HAVING, and QUALIFY, so I made a separate enum. 

A workaround could be:

* Extend the Clause Enum
* Keep it Private
* Create overloads like extendedExpandGroupBy, extendedExpandQualify, extendedExpandHaving ...

Reply:
It makes sense to add `QUALIFY` to `enum Clause`. It is a new SQL clause, after all.

If a method has a `Clause` parameter but does not accept all values, it can use `Preconditions.checkArgument` or just `throw` in the `default:` clause of a `switch`.

Are there any current or new uses outside the `org.apache.calcite.sql.validate` package? If not, there's no need for `public`.

Reply:
It get's used in GroupByScope which is still in the package, but outside the class. I can remove `public`, but I can't make it `private`

Reply:
Or I can make it private, but expose a public method like "extendExpandGroupBy"